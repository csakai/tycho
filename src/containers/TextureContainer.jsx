import React from 'react';
import {TextureLoader, FrontSide} from 'three';
import PropTypes from 'prop-types';
import Constants from '../constants';

export default class TextureContainer extends React.Component {

  static propTypes = {
      side: PropTypes.number,
      textures: PropTypes.array,
      transparent: PropTypes.bool
  }

  componentWillMount = () => {
      this.loadedTextures = [];
  }

  componentDidMount = () => {
      this.enqueueTextures(this.props.textures);
  }

  /**
   * Applies the given texture onto the material element.
   *
   * @param {Texture} texture - instance of THREE.Texture
   */
  onTextureLoaded = (textureData) => {
      this.loadedTextures.push(textureData);
  }

  /**
   * Enqueues the given texture set into the default THREE loader.
   *
   * @param {Object} prop
   * @param {String} prop.url - url of the texture to load
   */
  loadTexture = ({url, slot}) => {
      const loader = new TextureLoader();
      url = `/static/textures/map/${url}`; // TODO
      loader.load(url, () => this.onTextureLoaded({url, slot}));
  }

  /**
   * Calls loadTexture() on each texture set.
   *
   * @param {Object[]} textures - list of texture sets
   */
  enqueueTextures = (textures) => {
      if (Array.isArray(textures)) {
          textures.forEach(this.loadTexture);
      }
  }

  /**
   * Creates texture components from the loaded textures.
   *
   * @returns {Texture[]} array of texture components
   */
  getTextures = () => {
      return this
          .loadedTextures
          .map(({slot, url}, key) => {
              return <texture
                  url={url}
                  slot={slot}
                  key={key}
                  onLoad={this.updateMaterial}
              />
          });
  }

  /**
   * Flags material + assets to update on next animation frame.
   */
  updateMaterial = () => {
      const {material} = this.refs;

      if (material.map) {
          material.map.needsUpdate = true;
      }
      material.needsUpdate = true;
  }

  render() {
      return (
          <meshLambertMaterial
              color={Constants.WebGL.MESH_DEFAULT_COLOR}
              children={this.getTextures()}
              transparent={this.props.transparent}
              side={this.props.side || FrontSide}
              ref="material"
          />
      );
  }
}
