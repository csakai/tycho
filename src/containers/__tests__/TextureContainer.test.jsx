import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import TextureContainer from '../TextureContainer';

describe('Texture Container', () => {
    const textures = [
        {
            url: 'myImage.jpg',
            slot: 'map'
        }
    ];

    let component, textureContainer;

    beforeEach(() => {
        component = shallow(<TextureContainer
            textures={textures}
        />);
        textureContainer = component.instance();
    });

    describe('componentDidMount()', () => {
        it('should enqueue the given textures', () => {
            const spy = jest.spyOn(textureContainer, 'enqueueTextures');

            textureContainer.componentDidMount();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
            expect(spy).toHaveBeenCalledWith(textures);
        });
    });

    describe('onTextureLoaded()', () => {
        it('should push the given texture to the arrya of loaded textures', () => {
            const texture = textures[0];

            textureContainer.loadedTextures = [];
            textureContainer.onTextureLoaded(texture);
      
            expect(textureContainer.loadedTextures).toHaveLength(1);
            expect(textureContainer.loadedTextures[0]).toEqual(texture);
        });
    });

    describe('enqueueTextures()', () => {
        let spy;

        beforeEach(() => {
            spy = jest.spyOn(textureContainer, 'loadTexture');
        });

        it('should call loadTexture for each texture in the list', () => {
            textureContainer.enqueueTextures(textures);

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(textures.length);
        });

        it('should not call loadTexture if the param is not an array', () => {
            textureContainer.enqueueTextures();
      
            expect(spy).not.toHaveBeenCalled();
        });
    });

    describe('getTextures()', () => {
        let result;

        beforeEach(() => {
            textureContainer.loadedTextures = textures;
            result = textureContainer.getTextures();
        });

        it('should be an array having the same length of loadedTextures', () => {
            expect(result).toHaveLength(textures.length);
        });

        it('should set each entry to be a texture component', () => {
            expect(toJson(result)).toMatchSnapshot();
        });
    });

    describe('updateMaterial()', () => {
        beforeEach(() => {
            textureContainer.refs = {
                material: {
                    map: {}
                }
            };
        });

        it('should mark the map to require an update', () => {
            textureContainer.updateMaterial();
            expect(textureContainer.refs.material.needsUpdate).toEqual(true);
        });
    
        describe('when a map exists', () => {
            it('should mark the map to require an update', () => {
                textureContainer.updateMaterial();
                expect(textureContainer.refs.material.map.needsUpdate).toEqual(true);
            });
        });

        describe('when a map does not exist', () => {
            it('should have an undefined map', () => {
                textureContainer.refs.material.map = undefined;
                textureContainer.updateMaterial();
      
                expect(textureContainer.refs.material.map).not.toBeDefined();
            });
        });
    });

    describe('render()', () => {
        it('should render the texture container successfully', () => {
            expect(toJson(component)).toMatchSnapshot();
        });
    });
});
