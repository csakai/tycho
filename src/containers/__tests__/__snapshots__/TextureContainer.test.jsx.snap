// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Texture Container getTextures() should set each entry to be a texture component 1`] = `undefined`;

exports[`Texture Container render() should render the texture container successfully 1`] = `
<meshLambertMaterial
  color={16777215}
/>
`;
